pkg/operator/starter.go

```diff
@@ -0,0 +1,146 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/defaultstorageclass"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/snapshotcrd"
+	"github.com/openshift/cluster-storage-operator/pkg/operatorclient"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	opclient "github.com/openshift/client-go/operator/clientset/versioned"
+	opinformers "github.com/openshift/client-go/operator/informers/externalversions"
+
+	cfgclientset "github.com/openshift/client-go/config/clientset/versioned"
+	cfginformers "github.com/openshift/client-go/config/informers/externalversions"
+
+	apiextclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apiextinformers "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions"
+
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"github.com/openshift/library-go/pkg/operator/loglevel"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"github.com/openshift/library-go/pkg/operator/status"
+)
+
+const (
+	resync = 20 * time.Minute
+)
+
+const (
+	operatorNamespace   = "openshift-cluster-storage-operator"
+	clusterOperatorName = "storage"
+)
+
+func RunOperator(ctx context.Context, controllerConfig *controllercmd.ControllerContext) error {
+	kubeClient, err := kubernetes.NewForConfig(controllerConfig.ProtoKubeConfig)
+	if err != nil {
+		return err
+	}
+	kubeInformers := informers.NewSharedInformerFactory(kubeClient, resync)
+
+	operatorClientSet, err := opclient.NewForConfig(controllerConfig.KubeConfig)
+	if err != nil {
+		return err
+	}
+	operatorInformers := opinformers.NewSharedInformerFactoryWithOptions(operatorClientSet, resync,
+		opinformers.WithTweakListOptions(singleNameListOptions(operatorclient.GlobalConfigName)),
+	)
+
+	cfgClientset, err := cfgclientset.NewForConfig(controllerConfig.KubeConfig)
+	if err != nil {
+		return err
+	}
+	cfgInformers := cfginformers.NewSharedInformerFactoryWithOptions(cfgClientset, resync)
+
+	apiExtClientset, err := apiextclient.NewForConfig(controllerConfig.KubeConfig)
+	if err != nil {
+		return err
+	}
+	apiExtInformers := apiextinformers.NewSharedInformerFactoryWithOptions(apiExtClientset, resync)
+
+	operatorClient := &operatorclient.OperatorClient{
+		Informers: operatorInformers,
```

Comment:
So we are adding 4 differnet cliensets and it appears that `NewForConfig` itself does not document intended purpose. Can we document at each initialization point why we are initializing 4 different cliensets?

Reply:
Added a comment for each client